arch-install.bash

```diff
@@ -86,47 +90,106 @@ while true; do
   clear
 done
 
-sgdisk --zap-all "$block_device"
+if [[ "$encryption_choice" == "Yes" ]]; then
+  wipe_choice=$(printf 'Yes\nNo\n' | fzy -p 'do you want to secure wipe your disk?')
+  if [[ "$wipe_choice" == "Yes" ]]; then
+    drive_choice=$(printf 'SSD\nHDD\n' | fzy -p 'what kind of drive do you have?')
+    # Erase disk
+    case "$drive_choice" in 
+      #wont add nvme for now. No way to test
+      'SSD')
+        frozen_state=$(hdparm -I "$block_device" 2>/dev/null | awk '/frozen/ { print $1,$2 }')
+        if [ "${frozen_state}" == "not frozen" ]; then
+            hdparm --user-master u --security-set-pass password "$block_device"
+            hdparm --user-master u --security-erase password "$block_device"
+        else
+            echo "Your drive is frozen. Please fix!" >&2
+            exit 1 
+        fi
+      ;;
+      'HDD')
+        cryptsetup open --type plain -d /dev/urandom "$block_device" to_be_wiped
+        dd if=/dev/zero of=/dev/mapper/to_be_wiped status=progress bs=1M
+        cryptsetup close to_be_wiped
```

Comment:
This will fail with `No space left on device.` As far as I'm aware, there is no reason why this needs to be done via `cryptsetup` (please correct me if I'm wrong). We could easily replace these lines with `shred --random-source /dev/urandom --iterations 1 --zero --verbose "$block_device"`. This has the additional benefit of zeroing the drive after the random pass.

Reply:
https://wiki.archlinux.org/title/Dm-crypt/Drive_preparation#dm-crypt_wipe_on_an_empty_disk_or_partition

I followed the official wiki for this approach. I dont really know how exactly it failed for you in your VM as it worked in my VM and my X240 Laptop.

Could you maybe provide some kind of logs?

Also the wiping is not done via cryptsetup. It is done via the `dd` command.
the cryptsetup command only creates a temporary container 

The `-d` does this `-d, --key-file=STRING                 Read the key from a file` as per official arguments.

It is done that way to provide quote "very simple procedure to use an existing dm-crypt-volume to wipe all free space accessible on the underlying block device with random data by acting as a simple pseudorandom number generator. It is also claimed to protect against disclosure of usage patterns. That is because encrypted data is practically indistinguishable from random. "

Reply:
I just see what you meant. Yes the "No space left on device" is intentional.
Please read the wiki page for this. this is not an error but desired behaviour

Reply:
You're correct, but the problem remains: The script will exit as soon as any command fails. I suggest using `dd if=/dev/zero of=/dev/mapper/to_be_wiped status=progress bs=1M || true` to stop it from exiting the script. This is sort of a hack since it will ignore *all* errors instead of this specific one, but it's probably fine.

Reply:
If you want to do it in a more granular way, you can check whether the exit code is 28: `dd if=/dev/zero of=/dev/mapper/to_be_wiped status=progress bs=1M || if [[ "$?" != '28' ]]; then exit 1; fi`

Reply:
I does not make sense to capture a specific exit code as the exit code will be "1" on "No Space left on device".
This is proven through my testing. Therefore i will go with the generally accepted approach of just piping it to true

Reply:
Interesting. To clarify: It's not piping anything; it's just resetting the most recent exit status to 0 (success) in case `dd` fails.